packages/client/ui/react-ui/src/providers/CrossmintWalletProvider.tsx

```diff
@@ -74,61 +79,47 @@ export function CrossmintWalletProvider({
-            if (passkeySigner.onPrePasskeyRegistration == null) {
-                passkeySigner.onPrePasskeyRegistration = () =>
-                    new Promise<void>((resolve) => {
-                        setPasskeyPromptState({
-                            type: "create-wallet",
-                            open: true,
-                            primaryActionOnClick: () => {
-                                setPasskeyPromptState({ open: false });
-                                resolve();
-                            },
-                        });
-                    });
-            }
-
-            if (passkeySigner.onPasskeyRegistrationError == null) {
-                passkeySigner.onPasskeyRegistrationError = () => {
-                    return new Promise<void>((resolve) => {
-                        setPasskeyPromptState({
-                            type: "create-wallet-error",
-                            open: true,
-                            primaryActionOnClick: async () => {
-                                setPasskeyPromptState({ open: false });
-                                // When retrying, we don't want to show the passkey prompt again so unset the wallet creation callback
-                                passkeySigner.onPrePasskeyRegistration = undefined;
-                                await initializeWallet();
-                                resolve();
-                            },
-                        });
-                    });
-                };
-            }
+        try {
+            setWalletState({ status: "in-progress" });
+            const wallet = await smartWalletSDK.getOrCreateWallet(
+                { jwt: crossmint.jwt as string },
+                defaultChain,
+                enhanceConfigWithPasskeyPrompts(config)
+            );
+            setWalletState({ status: "loaded", wallet });
+        } catch (error: unknown) {
+            console.error("There was an error creating a wallet ", error);
+            setWalletState(deriveErrorState(error));
         }
-
-        const initializeWallet = async () => {
-            try {
-                setWalletState({ status: "in-progress" });
-                const wallet = await smartWalletSDK.getOrCreateWallet(
-                    { jwt: crossmint.jwt as string },
-                    defaultChain,
-                    config
-                );
-                setWalletState({ status: "loaded", wallet });
-            } catch (error: unknown) {
-                console.error("There was an error creating a wallet ", error);
-                setWalletState(deriveErrorState(error));
-            }
-        };
-
-        await initializeWallet();
-
         return { startedCreation: true };
     };
 
+    const enhanceConfigWithPasskeyPrompts = (config: WalletConfig) => {
+        if (enablePasskeyPrompt && config.signer && "type" in config.signer && config.signer.type === "PASSKEY") {
```

Comment:
This if is still a bit hard to read and follow. If we keep current type checking, I would suggest something like this:

```
const enhanceConfigWithPasskeyPrompts = (config: WalletConfig) => {
  const { signer } = config;

  if (!enablePasskeyPrompt || !signer || signer.type !== "PASSKEY") {
    return config;
  }

  return {
    ...config,
    signer: {
      ...signer,
      onPrePasskeyRegistration: createPasskeyPrompt("create-wallet"),
      onPasskeyRegistrationError: createPasskeyPrompt("create-wallet-error"),
    },
  };
};
```

Another alternative is to have stricter type checking:

```
type WalletConfigWithPasskeySigner = WalletParams & {
  signer: PasskeySigner;
  type: "evm-smart-wallet";
};
```

And use TS to test the type:

```
const isWalletConfigWithPasskeySigner = (
  config: WalletConfig
): config is WalletConfigWithPasskeySigner => config.signer.type === "PASSKEY"
```

Reply:
@AlbertoElias I've made this check a lot better imo, I think you'll agree 😄